### ISSUE TITLE
Document guarantees of our API calls

### ISSUE THREAD

**Issue by barakmich** [coreos/etcd#2668]
Especially as we design v3, but it'd be a helpful documentation bug for v2 as well.

When we have guarantees ("will return at this index/horizon", "will function atomically", etc) we should state them explicitly, every time.

**Comment by philips** [comment]
@barakmich Can you give me a concrete example of this?

**Comment by barakmich** [comment]
From the mailing list today: 

>Just wanted to make sure that I fully understand the consistency guarantees that etcd gives.  When I do a read or a watch, is it correct that all the values that I see are committed?  Ie. I might read stale data (if the server I'm reading from is behind) but I'll never read data that is then later determined to be wrong.  I think that's the case (and the promise of RAFT, as I understand it) but then someone asked me for a reference and I couldn't spot one in the API spec.

We should make sure that's clear in our docs per-function-call is what I'm suggesting, especially in clients/protos/etc, eg:
```
// GetKey returns the value for a key at the most recent fully-committed index.... blah...
// unless passed an Index parameter, in which case it will (a) block until that index has been committed // or (b) return the history at that index....
func (e *etcdclient) GetKey(...)
```

Likewise in API docs under /docs.

**Comment by xiang90** [comment]
closed via #4718
